File: web/src/data/FirestoreUsage.ts
/**
 * Bucketed counters for Firestore reads and writes, segmented by operation
 * name and event type. Mirrors Android's `FirestoreUsage.kt` so reports look
 * the same on both platforms — keep the data model and formatters in sync.
 *
 * Persistence: buckets are stored in localStorage. Writes are throttled to
 * once per 30s to keep the per-call hot path cheap; the latest unsaved
 * increment is lost on a hard tab-close (acceptable — diagnostic only).
 */

export type ReadType =
  | 'DOC_GET'
  | 'GET_DOCS'
  | 'LISTENER_INITIAL_FRESH'
  | 'LISTENER_UPDATE_FRESH'
  | 'LISTENER_INITIAL_CACHED'
  | 'LISTENER_UPDATE_CACHED'
  | 'LISTENER_LOCAL_ECHO'
  // Delta pull: bills exactly the docs returned (changed since lastSync).
  | 'PULL_DELTA'
  // Full repair pull triggered by count() mismatch — drops the watermark and
  // re-reads every doc.
  | 'PULL_FULL_REPAIR'
  // count() aggregation query used by foreground detection (~1 read).
  | 'COUNT_AGGREGATION'
  // Hydration of rawNotes from Firestore's IndexedDB cache on attach. $0.
  | 'HYDRATE_CACHED'

export type WriteType =
  | 'SET'
  | 'UPDATE'
  | 'DELETE'
  | 'BATCH_COMMIT'
  | 'TRANSACTION'

const LOCAL_ONLY_READ_TYPES = new Set<string>([
  'LISTENER_INITIAL_CACHED',
  'LISTENER_UPDATE_CACHED',
  'LISTENER_LOCAL_ECHO',
  'HYDRATE_CACHED',
])

interface Counter {
  ops: number
  docs: number
}

interface Bucket {
  start: number
  reads: Record<string, Counter>
  writes: Record<string, Counter>
}

const HOUR_MS = 60 * 60 * 1000
const DAY_MS = 24 * HOUR_MS
const MAX_HOURLY = 24
// Match the Firebase Console's monthly billing window so the report is
// directly comparable to the per-project usage shown in the console.
const MAX_DAILY = 30
const PERSIST_DEBOUNCE_MS = 30_000

const STORAGE_KEY_HOURLY = 'firestoreUsage:hourly'
const STORAGE_KEY_DAILY = 'firestoreUsage:daily'

class FirestoreUsage {
  private hourly: Bucket[] = []
  private daily: Bucket[] = []
  private persistTimer: ReturnType<typeof setTimeout> | null = null

  constructor() {
    this.load()
  }

  recordRead(operation: string, type: ReadType, docCount = 1): void {
    this.record(true, operation, type, docCount)
  }

  recordWrite(operation: string, type: WriteType, docCount = 1): void {
    this.record(false, operation, type, docCount)
  }

  private record(isRead: boolean, operation: string, type: string, docCount: number): void {
    const now = Date.now()
    const hour = this.currentBucket(this.hourly, hourStart(now), MAX_HOURLY)
    const day = this.currentBucket(this.daily, dayStart(now), MAX_DAILY)
    const key = `${operation}|${type}`
    incrementCounter(isRead ? hour.reads : hour.writes, key, docCount)
    incrementCounter(isRead ? day.reads : day.writes, key, docCount)
    this.schedulePersist()
  }

  private currentBucket(buckets: Bucket[], bucketStart: number, max: number): Bucket {
    const last = buckets[buckets.length - 1]
    if (last && last.start === bucketStart) return last
    const next: Bucket = { start: bucketStart, reads: {}, writes: {} }
    buckets.push(next)
    while (buckets.length > max) buckets.shift()
    return next
  }

  getReport(): string {
    const now = Date.now()
    const current = this.hourly[this.hourly.length - 1] ?? null
    const windows: [string, Bucket][] = [
      ['Last 24 hours', this.summarize(this.hourly.filter((b) => b.start >= now - 24 * HOUR_MS))],
      ['Last 7 days', this.summarize(this.daily.filter((b) => b.start >= now - 7 * DAY_MS))],
      ['Last 30 days (compare to Firebase console)', this.summarize(this.daily.filter((b) => b.start >= now - 30 * DAY_MS))],
    ]
    const lines: string[] = ['=== Firestore Usage Report ===']
    appendTimeSeries(lines, '== Billed ==', current, windows, formatBilledBucket)
    appendTimeSeries(lines, '== Local-only ==', current, windows, formatLocalBucket)
    lines.push('', '', '=== End ===')
    return lines.join('\n')
  }

  private summarize(buckets: Bucket[]): Bucket {
    const merged: Bucket = { start: 0, reads: {}, writes: {} }
    for (const b of buckets) {
      for (const [k, c] of Object.entries(b.reads)) accumulate(merged.reads, k, c)
      for (const [k, c] of Object.entries(b.writes)) accumulate(merged.writes, k, c)
    }
    return merged
  }

  reset(): void {
    this.hourly = []
    this.daily = []
    this.persist()
  }

  private load(): void {
    if (typeof localStorage === 'undefined') return
    try {
      const h = localStorage.getItem(STORAGE_KEY_HOURLY)
      const d = localStorage.getItem(STORAGE_KEY_DAILY)
      if (h) this.hourly = JSON.parse(h) as Bucket[]
      if (d) this.daily = JSON.parse(d) as Bucket[]
    } catch (e) {
      console.warn('FirestoreUsage: failed to load buckets from storage', e)
    }
  }

  private schedulePersist(): void {
    if (this.persistTimer != null) return
    this.persistTimer = setTimeout(() => {
      this.persistTimer = null
      this.persist()
    }, PERSIST_DEBOUNCE_MS)
  }

  private persist(): void {
    if (typeof localStorage === 'undefined') return
    try {
      localStorage.setItem(STORAGE_KEY_HOURLY, JSON.stringify(this.hourly))
      localStorage.setItem(STORAGE_KEY_DAILY, JSON.stringify(this.daily))
    } catch (e) {
      console.warn('FirestoreUsage: failed to persist buckets', e)
    }
  }
}

function hourStart(t: number): number {
  const d = new Date(t)
  d.setMinutes(0, 0, 0)
  return d.getTime()
}

function dayStart(t: number): number {
  const d = new Date(t)
  d.setHours(0, 0, 0, 0)
  return d.getTime()
}

function incrementCounter(map: Record<string, Counter>, key: string, docCount: number): void {
  const c = map[key]
  if (c) {
    c.ops += 1
    c.docs += docCount
  } else {
    map[key] = { ops: 1, docs: docCount }
  }
}

function accumulate(map: Record<string, Counter>, key: string, c: Counter): void {
  const existing = map[key]
  if (existing) {
    existing.ops += c.ops
    existing.docs += c.docs
  } else {
    map[key] = { ops: c.ops, docs: c.docs }
  }
}

const MONTH_ABBR = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

function formatHour(t: number): string {
  const d = new Date(t)
  const month = MONTH_ABBR[d.getMonth()]
  const day = d.getDate()
  const hour = String(d.getHours()).padStart(2, '0')
  return `${month} ${day} ${hour}:00`
}

function pluralize(noun: string, count: number): string {
  return count === 1 ? noun : `${noun}s`
}

function formatDocsOps(docs: number, ops: number): string {
  return `${docs} ${pluralize('doc', docs)} over ${ops} ${pluralize('op', ops)}`
}

function formatEntry(key: string, c: Counter): string {
  const sep = key.indexOf('|')
  const op = sep === -1 ? key : key.substring(0, sep)
  const rawType = sep === -1 ? '' : key.substring(sep + 1)
  const type = rawType.startsWith('LISTENER_') ? rawType.substring('LISTENER_'.length) : rawType
  return `${op} ${type}: ${formatDocsOps(c.docs, c.ops)}`
}

function isBilled(key: string): boolean {
  // Unknown types (including older lowercase keys still in localStorage)
  // default to billed so they surface in the report rather than disappearing.
  const type = key.substring(key.indexOf('|') + 1)
  return !LOCAL_ONLY_READ_TYPES.has(type)
}

function appendTimeSeries(
  lines: string[],
  sectionLabel: string,
  current: Bucket | null,
  windows: [string, Bucket][],
  formatter: (b: Bucket) => string,
): void {
  lines.push('', '', sectionLabel)
  if (current) {
    lines.push('', `Current hour (${formatHour(current.start)}):`, formatter(current))
  }
  for (const [label, bucket] of windows) {
    lines.push('', '', `${label}:`, formatter(bucket))
  }
}

function formatBilledBucket(b: Bucket): string {
  const billedReads = Object.entries(b.reads).filter(([key]) => isBilled(key))
  const totalReadOps = billedReads.reduce((sum, [, c]) => sum + c.ops, 0)
  const totalReadDocs = billedReads.reduce((sum, [, c]) => sum + c.docs, 0)
  const writeEntries = Object.entries(b.writes)
  const totalWriteOps = writeEntries.reduce((sum, [, c]) => sum + c.ops, 0)
  const totalWriteDocs = writeEntries.reduce((sum, [, c]) => sum + c.docs, 0)
  const ratioSuffix = totalWriteDocs > 0
    ? ` (R:W ${(totalReadDocs / totalWriteDocs).toFixed(1)}x)`
    : ''
  const lines: string[] = []
  lines.push(`  Reads: ${formatDocsOps(totalReadDocs, totalReadOps)}`)
  for (const [key, c] of billedReads.sort((a, b) => b[1].docs - a[1].docs)) {
    lines.push(`    ${formatEntry(key, c)}`)
  }
  lines.push(`  Writes: ${formatDocsOps(totalWriteDocs, totalWriteOps)}${ratioSuffix}`)
  for (const [key, c] of writeEntries.sort((a, b) => b[1].docs - a[1].docs)) {
    lines.push(`    ${formatEntry(key, c)}`)
  }
  return lines.join('\n')
}

function formatLocalBucket(b: Bucket): string {
  const localReads = Object.entries(b.reads).filter(([key]) => !isBilled(key))
  const totalDocs = localReads.reduce((sum, [, c]) => sum + c.docs, 0)
  const totalOps = localReads.reduce((sum, [, c]) => sum + c.ops, 0)
  const lines: string[] = []
  lines.push(`  Reads: ${formatDocsOps(totalDocs, totalOps)}`)
  for (const [key, c] of localReads.sort((a, b) => b[1].docs - a[1].docs)) {
    lines.push(`    ${formatEntry(key, c)}`)
  }
  return lines.join('\n')
}

export const firestoreUsage = new FirestoreUsage()

if (typeof window !== 'undefined') {
  ;(window as unknown as { firestoreUsage: FirestoreUsage }).firestoreUsage = firestoreUsage
}

File: app/src/main/java/org/alkaline/taskbrain/data/FirestoreUsage.kt
package org.alkaline.taskbrain.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.util.Calendar
import java.util.concurrent.TimeUnit

/**
 * Bucketed counters for Firestore reads and writes, segmented by operation
 * name and event type. Used to find hotspots — call [recordRead] /
 * [recordWrite] at every Firestore call site, then tap the Usage button on
 * the note list to view the report.
 *
 * Data model:
 * - Hourly buckets: last 24 hours, one bucket per local-clock hour.
 * - Daily buckets: last 7 days, one bucket per local-clock day.
 * - Each bucket holds (operation, type) → {ops, docs} for both reads and writes.
 *
 * Persistence: buckets are stored in SharedPreferences and survive app
 * restarts. Writes are debounced to once per 30s to keep the hot path cheap;
 * the latest unsaved increment is lost on a hard kill (acceptable — this is
 * diagnostic data, not user data). Call [attach] once from
 * `Application.onCreate` so reads/writes have somewhere to persist.
 */
object FirestoreUsage {
    private const val TAG = "FirestoreUsage"
    private const val PREFS_NAME = "firestore_usage"
    private const val KEY_HOURLY = "hourly"
    private const val KEY_DAILY = "daily"
    private val HOUR_MS = TimeUnit.HOURS.toMillis(1)
    private val DAY_MS = TimeUnit.DAYS.toMillis(1)
    private const val MAX_HOURLY = 24
    private const val MAX_DAILY = 7
    private const val PERSIST_DEBOUNCE_MS = 30_000L

    enum class ReadType {
        DOC_GET,
        GET_DOCS,
        LISTENER_INITIAL_FRESH,
        LISTENER_INITIAL_CACHED,
        LISTENER_UPDATE_FRESH,
        LISTENER_UPDATE_CACHED,
        LISTENER_LOCAL_ECHO,
    }

    enum class WriteType {
        SET, UPDATE, DELETE, BATCH_COMMIT, TRANSACTION,
    }

    private data class Counter(var ops: Long = 0, var docs: Long = 0)

    private data class Bucket(
        val start: Long,
        val reads: HashMap<String, Counter> = HashMap(),
        val writes: HashMap<String, Counter> = HashMap(),
    )

    private val hourly = ArrayList<Bucket>()
    private val daily = ArrayList<Bucket>()
    private val lock = Any()

    private var prefs: SharedPreferences? = null
    private val persistHandler = android.os.Handler(android.os.Looper.getMainLooper())
    private var persistScheduled = false

    /** Hook up persistence. Idempotent; safe to call from Application.onCreate. */
    fun attach(context: Context) {
        synchronized(lock) {
            if (prefs != null) return
            prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            load()
        }
    }

    fun recordRead(operation: String, type: ReadType, docCount: Int = 1) {
        record(isRead = true, operation = operation, type = type.name, docCount = docCount)
    }

    fun recordWrite(operation: String, type: WriteType, docCount: Int = 1) {
        record(isRead = false, operation = operation, type = type.name, docCount = docCount)
    }

    private fun record(isRead: Boolean, operation: String, type: String, docCount: Int) {
        val now = System.currentTimeMillis()
        synchronized(lock) {
            val hour = currentBucket(hourly, hourStart(now), MAX_HOURLY)
            val day = currentBucket(daily, dayStart(now), MAX_DAILY)
            val key = "$operation|$type"
            increment(if (isRead) hour.reads else hour.writes, key, docCount)
            increment(if (isRead) day.reads else day.writes, key, docCount)
            schedulePersist()
        }
    }

    private fun currentBucket(buckets: ArrayList<Bucket>, bucketStart: Long, max: Int): Bucket {
        val last = buckets.lastOrNull()
        if (last != null && last.start == bucketStart) return last
        val next = Bucket(bucketStart)
        buckets.add(next)
        while (buckets.size > max) buckets.removeAt(0)
        return next
    }

    fun getReport(): String = synchronized(lock) {
        buildString {
            appendLine("=== Firestore Usage Report ===")
            val now = System.currentTimeMillis()
            val current = hourly.lastOrNull()
            if (current != null) {
                appendLine()
                appendLine("Current hour (${formatHour(current.start)}):")
                append(formatBucket(current))
            }
            val last24 = summarize(hourly.filter { it.start >= now - 24 * HOUR_MS })
            appendLine()
            appendLine()
            appendLine("Last 24 hours:")
            append(formatBucket(last24))
            val last7 = summarize(daily.filter { it.start >= now - 7 * DAY_MS })
            appendLine()
            appendLine()
            appendLine("Last 7 days:")
            append(formatBucket(last7))
            appendLine()
            appendLine()
            append("=== End ===")
        }
    }

    fun reset() {
        synchronized(lock) {
            hourly.clear()
            daily.clear()
            persist()
        }
    }

    private fun summarize(buckets: List<Bucket>): Bucket {
        val merged = Bucket(0)
        for (b in buckets) {
            for ((k, c) in b.reads) accumulate(merged.reads, k, c)
            for ((k, c) in b.writes) accumulate(merged.writes, k, c)
        }
        return merged
    }

    private fun increment(map: HashMap<String, Counter>, key: String, docCount: Int) {
        val c = map.getOrPut(key) { Counter() }
        c.ops += 1
        c.docs += docCount
    }

    private fun accumulate(map: HashMap<String, Counter>, key: String, c: Counter) {
        val existing = map.getOrPut(key) { Counter() }
        existing.ops += c.ops
        existing.docs += c.docs
    }

    private fun hourStart(t: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = t; set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0) }
        return cal.timeInMillis
    }

    private fun dayStart(t: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = t; set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0) }
        return cal.timeInMillis
    }

    private fun formatHour(t: Long): String {
        val cal = Calendar.getInstance().apply { timeInMillis = t }
        return "%tb %td %02d:00".format(cal, cal, cal.get(Calendar.HOUR_OF_DAY))
    }

    private fun formatBucket(b: Bucket): String = buildString {
        var totalReadOps = 0L
        var totalReadDocs = 0L
        var totalWriteOps = 0L
        var totalWriteDocs = 0L
        appendLine("  Reads:")
        for ((key, c) in b.reads.entries.sortedByDescending { it.value.docs }) {
            val parts = key.split("|", limit = 2)
            val op = parts.getOrNull(0) ?: ""
            val type = parts.getOrNull(1) ?: ""
            appendLine("    ${op.padEnd(34)} ${type.padEnd(28)} ${"%5d".format(c.ops)} ops × ${"%6d".format(c.docs)} docs")
            totalReadOps += c.ops
            totalReadDocs += c.docs
        }
        appendLine("    TOTAL: $totalReadOps ops × $totalReadDocs docs")
        appendLine("  Writes:")
        for ((key, c) in b.writes.entries.sortedByDescending { it.value.docs }) {
            val parts = key.split("|", limit = 2)
            val op = parts.getOrNull(0) ?: ""
            val type = parts.getOrNull(1) ?: ""
            appendLine("    ${op.padEnd(34)} ${type.padEnd(28)} ${"%5d".format(c.ops)} ops × ${"%6d".format(c.docs)} docs")
            totalWriteOps += c.ops
            totalWriteDocs += c.docs
        }
        append("    TOTAL: $totalWriteOps ops × $totalWriteDocs docs")
        if (totalWriteDocs > 0) {
            appendLine()
            append("    R:W RATIO: %.1fx".format(totalReadDocs.toDouble() / totalWriteDocs))
        }
    }

    private fun load() {
        val p = prefs ?: return
        try {
            p.getString(KEY_HOURLY, null)?.let { hourly.addAll(parseBucketList(it)) }
            p.getString(KEY_DAILY, null)?.let { daily.addAll(parseBucketList(it)) }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to load buckets from prefs; resetting", e)
            hourly.clear()
            daily.clear()
        }
    }

    private fun schedulePersist() {
        if (persistScheduled) return
        persistScheduled = true
        persistHandler.postDelayed({
            synchronized(lock) {
                persistScheduled = false
                persist()
            }
        }, PERSIST_DEBOUNCE_MS)
    }

    private fun persist() {
        val p = prefs ?: return
        try {
            p.edit()
                .putString(KEY_HOURLY, serializeBucketList(hourly))
                .putString(KEY_DAILY, serializeBucketList(daily))
                .apply()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to persist buckets", e)
        }
    }

    private fun serializeBucketList(buckets: List<Bucket>): String {
        val arr = JSONArray()
        for (b in buckets) {
            val obj = JSONObject()
            obj.put("start", b.start)
            obj.put("reads", serializeCounters(b.reads))
            obj.put("writes", serializeCounters(b.writes))
            arr.put(obj)
        }
        return arr.toString()
    }

    private fun serializeCounters(map: Map<String, Counter>): JSONObject {
        val obj = JSONObject()
        for ((k, c) in map) {
            obj.put(k, JSONArray().apply { put(c.ops); put(c.docs) })
        }
        return obj
    }

    private fun parseBucketList(json: String): List<Bucket> {
        val arr = JSONArray(json)
        val out = ArrayList<Bucket>(arr.length())
        for (i in 0 until arr.length()) {
            val obj = arr.getJSONObject(i)
            val b = Bucket(obj.getLong("start"))
            b.reads.putAll(parseCounters(obj.getJSONObject("reads")))
            b.writes.putAll(parseCounters(obj.getJSONObject("writes")))
            out.add(b)
        }
        return out
    }

    private fun parseCounters(obj: JSONObject): Map<String, Counter> {
        val out = HashMap<String, Counter>(obj.length())
        val keys = obj.keys()
        while (keys.hasNext()) {
            val k = keys.next()
            val arr = obj.getJSONArray(k)
            out[k] = Counter(ops = arr.getLong(0), docs = arr.getLong(1))
        }
        return out
    }
}
